logging: let docker-py handle utf8 and json decode

jarodwilson opened the pull request:
Rather than always calling .decode("utf-8") within CommandResult's
parse_item function and then interpreting the data as json, set
decode=True in callers to docker.Client.build, so decoding and de-json'ing
happens in docker-py, and we just get a nice clean generator full of
dicts.

Note: this also requires updating the inputs for
test_util.py::test_parse_item, since decoding is happening externally.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

jarodwilson commented:
py.test tests/test_util.py passes on both an el7 host with python 2.7 and docker 1.x and a fedora 26 host with python 3.6 and docker 2.4.x (with some modifications for Client->APIClient).

twaugh commented:
There are some more tests which assume the old behaviour:
```
$ git grep -l "b'{\""
tests/docker_mock.py
tests/plugins/test_tag_and_push.py
```

jarodwilson commented:
All self-tests, not just test_util.py, are now passing, discovered a minor error in the earlier error logging code and fixed it (can't call .strip() on a dict), fixed up docker_mock.py and test_tag_and_push.py.

jarodwilson commented:
Argh. Now I've got 15 test failures on python 3.x, on to hunting those...

jarodwilson commented:
The latest push includes a fixup for core.py's logs() function, which properly handles non-stream logs() calls under python3 now, which got all self-tests passing in travis on python2 and python3.

vrutkovs commented:
This causes a weird output in the logs from docker about progress details:
```
2017-08-01 09:19:31,333 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,084 platform:- - atomic_reactor.plugin - DEBUG - running plugin 'tag_and_push'
2017-08-01 09:19:31,340 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,084 platform:- - atomic_reactor.plugin - INFO - running plugin instance with args: '{u'registries': {u'172.17.0.1:5000': {u'insecure': True, u'secret': u'/var/run/secrets/atomic-reactor/v2-registry-dockercfg'}}}'
2017-08-01 09:19:31,343 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,084 platform:- - atomic_reactor.plugins.tag_and_push - INFO - Registry 172.17.0.1:5000 secret /var/run/secrets/atomic-reactor/v2-registry-dockercfg
2017-08-01 09:19:31,347 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,084 platform:- - atomic_reactor.core - INFO - tagging and pushing image 'f33a7b8fad513f560e8e88d554837e3f675de77ecc8da77ce9900ead0d34f4d0' as '172.17.0.1:5000/lucarval/docker-hello-world:1.0-2'
2017-08-01 09:19:31,350 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,084 platform:- - atomic_reactor.core - DEBUG - image = 'f33a7b8fad513f560e8e88d554837e3f675de77ecc8da77ce9900ead0d34f4d0', target_image = '172.17.0.1:5000/lucarval/docker-hello-world:1.0-2'
2017-08-01 09:19:31,352 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,084 platform:- - atomic_reactor.core - INFO - tagging image 'f33a7b8fad513f560e8e88d554837e3f675de77ecc8da77ce9900ead0d34f4d0' as '172.17.0.1:5000/lucarval/docker-hello-world:1.0-2'
2017-08-01 09:19:31,355 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,084 platform:- - atomic_reactor.core - DEBUG - image = 'f33a7b8fad513f560e8e88d554837e3f675de77ecc8da77ce9900ead0d34f4d0', target_image_name = '172.17.0.1:5000/lucarval/docker-hello-world:1.0-2'
2017-08-01 09:19:31,358 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,109 platform:- - atomic_reactor.core - INFO - logging in: registry '172.17.0.1:5000', secret path '/var/run/secrets/atomic-reactor/v2-registry-dockercfg'
2017-08-01 09:19:31,359 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,109 platform:- - atomic_reactor.core - INFO - found username osbs for registry 172.17.0.1:5000
2017-08-01 09:19:31,361 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,109 platform:- - atomic_reactor.core - INFO - pushing image '172.17.0.1:5000/lucarval/docker-hello-world:1.0-2'
2017-08-01 09:19:31,363 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,109 platform:- - atomic_reactor.core - DEBUG - image: '172.17.0.1:5000/lucarval/docker-hello-world:1.0-2', insecure: 'True'
2017-08-01 09:19:31,364 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,438 platform:- - atomic_reactor.util - INFO - wait_for_command
2017-08-01 09:19:31,367 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,438 platform:- - atomic_reactor.util - DEBUG - {"status":"The push refers to a repository [172.17.0.1:5000/lucarval/docker-hello-world]"}
2017-08-01 09:19:31,368 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,461 platform:- - atomic_reactor.util - DEBUG - {"status":"Preparing","progressDetail":{},"id":"5b47abb83b07"}
2017-08-01 09:19:31,370 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,461 platform:- - atomic_reactor.util - DEBUG - {"status":"Preparing","progressDetail":{},"id":"42ce6d637cc2"}
2017-08-01 09:19:31,372 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,489 platform:- - atomic_reactor.util - DEBUG - {"status":"Layer already exists","progressDetail":{},"id":"42ce6d637cc2"}
2017-08-01 09:19:31,375 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,556 platform:- - atomic_reactor.util - DEBUG - {"status":"Pushing","progressDetail":{"current":512,"total":634},"progress":"[========================================\u003e          ]    512 B/634 B","id":"5b47abb83b07"}
2017-08-01 09:19:31,379 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,560 platform:- - atomic_reactor.util - DEBUG - {"status":"Pushing","progressDetail":{"current":10240,"total":634},"progress":"[==================================================\u003e] 10.24 kB","id":"5b47abb83b07"}
2017-08-01 09:19:31,381 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:30,757 platform:- - atomic_reactor.util - DEBUG - {"status":"Pushed","progressDetail":{},"id":"5b47abb83b07"}
2017-08-01 09:19:31,383 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:31,196 platform:- - atomic_reactor.util - DEBUG - {"status":"1.0-2: digest: sha256:119e5decd64a88d5a7f2893a108cf646199b59d6028ad5123f724b59f7c00763 size: 736"}
2017-08-01 09:19:31,387 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:31,212 platform:- - atomic_reactor.util - DEBUG - {"progressDetail":{},"aux":{"Tag":"1.0-2","Digest":"sha256:119e5decd64a88d5a7f2893a108cf646199b59d6028ad5123f724b59f7c00763","Size":736}}
2017-08-01 09:19:31,388 platform:x86_64 - atomic_reactor.plugins.orchestrate_build - INFO - 2017-08-01 09:19:31,213 platform:- - atomic_reactor.util - INFO - no more logs
```

Not sure if this is expected

twaugh commented:
@vrutkovs, good catch. I think this is because there are other uses of `wait_for_command` which need to be changed to use `decode=True`:
* `DockerTasker.pull_image`
* `DockerTasker.push_image`
* `DockerTasker.build_image_from_path`
* `DockerTasker.build_image_from_git`
* but not `DockerTasker.logs` as `docker.APIClient.logs` takes no `decode` parameter

jarodwilson commented:
I believe build_image_from_* are both already covered by decode=True added to the d.build() call in build_image_from_path, but pull_image and push_image do both need decode=True added to their d.pull and d.push calls respectively, as well as needing to add that to some of the tests that call d.pull as well. Testing fixes locally now...

jarodwilson commented on tests/test_tasker.py:
```diff
@@ -37,7 +37,7 @@ def setup_module(module):
         setattr(module, 'HAS_IMAGE', True)
     except docker.errors.APIError:
-        _ = [x for x in d.pull(INPUT_IMAGE, stream=True)]
+        _ = [x for x in d.pull(INPUT_IMAGE, decode=True, stream=True)]
```
This is of course seems to be intentional, and isn't something new introduced by this patch.

lcarva replied:
It's safe to remove `_ = `
It was probably added prior to stickler being around.

twaugh commented:
Looks great! I'm happy once stickler-ci's happy.

lcarva commented:
@jarodwilson "This branch has conflicts that must be resolved"

twaugh commented:
Sorry, that was my fault. Just change `docker.Client` to `docker.APIClient`.

jarodwilson commented:
Rebased atop current master with Client->APIClient changes.

twaugh commented:
Draft release notes updated.
